executor/analyze.go

```diff
@@ -381,3 +398,350 @@ func (e *AnalyzeColumnsExec) buildStats() (hists []*statistics.Histogram, cms []
+	}
+	return result
+}
+
+// AnalyzeFastTask is the task for build stats.
+type AnalyzeFastTask struct {
+	Location  *tikv.KeyLocation
+	SampSize  uint64
+	LRowCount uint64
+	RRowCount uint64
+}
+
+// AnalyzeFastExec represents Fast Analyze Columns executor.
+type AnalyzeFastExec struct {
+	ctx             sessionctx.Context
+	PhysicalTableID int64
+	pkInfo          *model.ColumnInfo
+	colsInfo        []*model.ColumnInfo
+	idxsInfo        []*model.IndexInfo
+	concurrency     int
+	maxNumBuckets   uint64
+	table           table.Table
+	cache           *tikv.RegionCache
+	wg              *sync.WaitGroup
+	sampLocs        chan *tikv.KeyLocation
+	sampLocRowCount uint64
+	tasks           chan *AnalyzeFastTask
+	scanTasks       []*tikv.KeyLocation
+}
+
+func (e *AnalyzeFastExec) getSampRegionsRowCount(bo *tikv.Backoffer, needRebuild *bool, err *error) {
+	defer func() {
+		e.wg.Done()
+		if *needRebuild == true {
+			for _, ok := <-e.sampLocs; ok; _, ok = <-e.sampLocs {
+				// do nothing
+			}
+		}
+	}()
+	client := e.ctx.GetStore().(tikv.Storage).GetTiKVClient()
+	for {
+		loc, ok := <-e.sampLocs
+		if !ok {
+			return
+		}
+		req := &tikvrpc.Request{
+			Type: tikvrpc.CmdDebugGetRegionProperties,
+			DebugGetRegionProperties: &debugpb.GetRegionPropertiesRequest{
+				RegionId: loc.Region.GetID(),
+			},
+		}
+		var resp *tikvrpc.Response
+		var rpcCtx *tikv.RPCContext
+		rpcCtx, *err = e.cache.GetRPCContext(bo, loc.Region)
+		if *err != nil {
+			*err = errors.Trace(*err)
+			return
+		}
+		ctx := context.Background()
+		resp, *err = client.SendRequest(ctx, rpcCtx.Addr, req, tikv.ReadTimeoutMedium)
+		if *err != nil {
+			*err = errors.Trace(*err)
+			return
+		}
+		// TODO: check the region not_found
+		if ctx.Err() != nil {
+			fmt.Println(ctx.Err().Error())
+			*needRebuild = true
+			return
+		}
+		for _, prop := range resp.DebugGetRegionProperties.Props {
+			if prop.Name == "num_rows" {
+				var cnt uint64
+				cnt, *err = strconv.ParseUint(prop.Value, 10, 64)
+				if *err != nil {
+					*err = errors.Trace(*err)
+					return
+				}
+				newCount := atomic.AddUint64(&e.sampLocRowCount, cnt)
+				task := &AnalyzeFastTask{
+					Location:  loc,
+					LRowCount: newCount - cnt,
+					RRowCount: newCount,
+				}
+				e.tasks <- task
+			}
+		}
+	}
+}
+
+func (e *AnalyzeFastExec) buildSampTask() (bool, error) {
+	if e.wg == nil {
+		e.wg = &sync.WaitGroup{}
+	}
+	e.wg.Wait()
+
+	store, _ := e.ctx.GetStore().(tikv.Storage)
+	e.cache = store.GetRegionCache()
+	startKey, endKey := tablecodec.GetTableHandleKeyRange(e.table.Meta().ID)
+	var loc *tikv.KeyLocation
+	var err error
+	bo := tikv.NewBackoffer(context.Background(), 500)
+	for loc, err = e.cache.LocateKey(bo, startKey); bytes.Compare(loc.StartKey, endKey) <= 0 && err == nil; loc, err = e.cache.LocateKey(bo, append(loc.EndKey, byte(0))) {
+		if bytes.Compare(endKey, loc.EndKey) < 0 || bytes.Compare(loc.StartKey, startKey) < 0 {
+			e.scanTasks = append(e.scanTasks, loc)
```

Comment:
It seems that there is no other place using this `scanTasks`.

Reply:
It effects now.